packages/block-editor/src/components/list-view/use-block-selection.js

```diff
@@ -0,0 +1,129 @@
+/**
+ * WordPress dependencies
+ */
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useCallback } from '@wordpress/element';
+import { UP, DOWN } from '@wordpress/keycodes';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../../store';
+import { getCommonDepthClientIds } from './utils';
+
+export default function useBlockSelection() {
+	const { clearSelectedBlock, multiSelect, selectBlock } = useDispatch(
+		blockEditorStore
+	);
+	const {
+		getBlockParents,
+		getBlockSelectionStart,
+		getBlockSelectionEnd,
+		getNextBlockClientId,
+		getPreviousBlockClientId,
+		getSelectedBlockClientIds,
+		hasMultiSelection,
+		hasSelectedBlock,
+	} = useSelect( blockEditorStore );
+
+	const updateBlockSelection = useCallback(
+		async ( clientId, event ) => {
+			if ( ! event.shiftKey ) {
+				await clearSelectedBlock();
+				selectBlock( clientId );
+			} else if ( event.shiftKey ) {
+				// To handle multiple block selection via the `SHIFT` key, prevent
+				// the browser default behavior of opening the link in a new window.
+				event.preventDefault();
+
+				const isKeyPress =
+					event.type === 'keydown' &&
+					( event.keyCode === UP || event.keyCode === DOWN );
+				const isUp = event.keyCode === UP;
+				const isDown = event.keyCode === DOWN;
+				const selectedBlocks = getSelectedBlockClientIds();
+				const clientIdWithParents = [
+					...getBlockParents( clientId ),
+					clientId,
+				];
+
+				if (
+					isKeyPress &&
+					! selectedBlocks.some( ( blockId ) =>
+						clientIdWithParents.includes( blockId )
+					)
+				) {
+					// Ensure that shift-selecting blocks via the keyboard only
+					// expands the current selection if hovering over already
+					// selected blocks. Otherwise, clear the selection so that
+					// a user can create a new selection entirely by keyboard.
+					await clearSelectedBlock();
+				}
+
+				let startTarget = getBlockSelectionStart();
+				let endTarget = clientId;
+
+				// Handle clicking on a block when no blocks are selected.
+				if (
+					! isKeyPress &&
+					! hasSelectedBlock() &&
+					! hasMultiSelection()
+				) {
+					selectBlock( clientId, null );
+					return;
+				}
+
+				// Handle keyboard behavior for selecting multiple blocks.
```

Comment:
> I think generally my expectation would be that multi-selection follows where my focus is or vice versa

Great catch @talldan, thanks for the detailed look here, and for the suggestions. I tried a couple of things out and settled on deferring how we get the next/prev block id to the `TreeGrid` component by adding in an additional callback `onChangeRow` — this passes up the event object, the originally focused row, and the target focused row, so that we can then use them in `updateBlockSelection`. It also allows us to slightly simplify the logic in the `use-block-selection` hook.

Since we already pass row data in a similar way in `onExpandRow`, it seems to have a pretty good parallel there, while not introducing additional complexity to the `TreeGrid` component directly. Here's how it's looking now (as we press DOWN, the selection and focus moves to the child blocks of Navigation instead of to the next Group block):

![Kapture 2022-02-14 at 17 31 42](https://user-images.githubusercontent.com/14988353/153812508-6fac01ac-c789-493a-b794-8316b6124482.gif)

> In the main block list when I try this, my focus is constrained to only blocks at the same level, and I think this might also be the best way for it to work in List View.

Good point about the constrained focus in the editor 🤔. I'm not sure I agree that the same UX should apply in the list view, though. Because of the free and visible navigation within the tree view, I don't think holding shift should prevent the movement of which element we're focusing. My preference would be for us to keep the click-behaviour multi-select logic so that moving up a level automatically expands the selection to be more broad. E.g. from a deeply nested part of the hierarchy, I think it should be possible to press SHIFT+UP to select a broader region. But this could just be a personal preference 🙂!

Reply:
Sounds great, the gif looks really good.

Thanks for working on that solution. I'll test it out a bit later.

Reply:
Had a chance to test it now, and it's really good, great work! I'll leave some general comments on the main discussion.